crypto/evp/s_lib.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+#include <openssl/param_build.h>
+#include <openssl/evp.h>
+#include <openssl/core_names.h>
+
+#include "internal/provider.h"
+#include "crypto/evp.h"
+#include "evp_local.h"
+
+int EVP_SKEY_export(const EVP_SKEY *skey, int selection,
+                    OSSL_CALLBACK *export_cb, void *export_cbarg)
+{
+    if (skey == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    if (skey->skeymgmt == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+
+    return evp_skeymgmt_export(skey->skeymgmt, skey->keydata, selection, export_cb, export_cbarg);
+}
+
+EVP_SKEY *evp_skey_int(void)
+{
+    EVP_SKEY *skey = OPENSSL_zalloc(sizeof(EVP_SKEY));
+
+    if (!CRYPTO_NEW_REF(&skey->references, 1))
+        goto err;
+
+    skey->lock = CRYPTO_THREAD_lock_new();
+    if (skey->lock == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_CRYPTO_LIB);
+        goto err;
+    }
+    return skey;
+
+ err:
+    CRYPTO_FREE_REF(&skey->references);
+    CRYPTO_THREAD_lock_free(skey->lock);
+    OPENSSL_free(skey);
+    return NULL;
+}
+
+EVP_SKEY *EVP_SKEY_import(OSSL_LIB_CTX *libctx, const char *skeymgmtname, const char *propquery,
+                          int selection, const OSSL_PARAM *params)
+{
+    EVP_SKEYMGMT *skeymgmt = NULL;
+    EVP_SKEY *skey = evp_skey_int();
+
+    if (skey == NULL)
+        return NULL;
+
+    skeymgmt = EVP_SKEYMGMT_fetch(libctx, skeymgmtname, propquery);
+    if (skeymgmt == NULL) {
+        /* if the specific key_type is unkown, attempt to use the generic
+         * key management */
+        skeymgmt = EVP_SKEYMGMT_fetch(libctx, "GENERIC_SECRET", propquery);
```

Comment:
Should be a global define

Reply:
You mean "GENERIC_SECRET" ?
Suggestions on where to add it as global ?

Reply:
done